File: Python/Medium/backtracking/permutations.py
class Solution:
    def permute(self, nums):
        result = []
        visited = set()

        def backtrack(result, visited, subset, nums):
            if len(subset) == len(nums):
                result.append(subset.copy())

            for i in range(len(nums)):
                if i not in visited:
                    visited.add(i)
                    backtrack(result, visited, subset+[nums[i]], nums)
                    visited.remove(i)
        backtrack(result, visited, [], nums)
        return result

File: JavaScript/Medium/backtracking/permutations.js
/**
 * @param {number []} nums
 * @return {number [][]}
 */

Set()

var permute = function (nums) {
    let result = []
    let visited = new Set()

    var backtrack = (result, visited, subset, nums) => {
        if (subset.length === nums.length){
            result.push([...subset])
        }

        for (let i = 0; i<nums.length; i++) {
            if (!visited.has(i)) {
                visited.add(i)
                backtrack(result, visited, subset, nums)
                visited.remove(i)
            }
        }
    }

    

}